feat: adds creation, saving and playback of audio

cimigree opened the pull request:
closes #360 

### Description
Adds the ability to create an audio file from an observation
Adds the necessary files and logic so that when you hit "stop" on an audio recording, you are given the Playback option to be able to hear what you just recorded. 
Continues to address issue https://github.com/digidem/comapeo-mobile/issues/360
Used this PR that [Andrew created](https://github.com/digidem/comapeo-mobile/compare/develop...360/create-audio-recording)
Here are the [designs](https://www.figma.com/design/1Knqv1NrijUuqMo1pvYyF7/%22Ready-for-Dev%22-Hand-off-Assets?node-id=0-1&node-type=canvas&t=DhW0UI0lwgUo0EC5-0)
Uses some code from [SWM](https://github.com/digidem/comapeo-mobile/pull/381)

Instructions
Make sure to add this to your env file: EXPO_PUBLIC_FEATURE_AUDIO=true
Go to create an observation
Hit the microphone
Grant permission to record audio
You will be taken to the create audio recording screen
Watch the screen change color and the time pass and the countdown change.
It will stop automatically at 5 minutes
Hit the stop button to stop recording.
You can play, stop, play, stop, and then when you get to the end it resets to the beginning. Hope that is right. Designs weren't clear.
At that point, the audio is added to state. 
You can delete it with the trashcan.
If you hit the X in the top left, the audio is added to the draft observation, the success modal opens, and you are given the option to return to the editor or record another audio, which I wired up to work.
Once you save the Observation, the audio file or files is added to the observation.
You can't see these yet... That will be another PR!

cimigree committed:
Adds functionality for asking for permissions for Audio. Adds Audio navigation and dummy screen.

cimigree committed:
Adds messages

cimigree committed:
Adds ability to create a recording.

cimigree committed:
Adds messages.

cimigree committed:
Adds ability to playback and save audio recordings.

cimigree committed:
Adds messages.

cimigree committed:
Adds play arrow image.

cimigree committed:
Playback working and can play from midway.

cimigree committed:
Tightens playback.

cimigree committed:
Merges with develop.

cimigree committed:
Merge branch 'feat/add-audio-creation' into feat/add-audio-save-playback

ErikSin commented on src/frontend/screens/Audio/CreateRecording/RecordingDone.tsx:
```diff
@@ -0,0 +1,149 @@
+        />
+      ),
+    });
+  }, [navigation, openSuccessSheet]);
```
this is missing `addAudio`, `createdAt`, `duration`, and `uri` from the dependency array

ErikSin commented on src/frontend/screens/Audio/CreateRecording/index.tsx:
We should [prevent the user from going back](https://reactnavigation.org/docs/preventing-going-back) on this page (except for when `recording.staus === idle`)

ErikSin commented on src/frontend/screens/Audio/CreateRecording/RecordingIdle.tsx:
```diff
@@ -0,0 +1,46 @@
+  const navigation = useNavigationFromRoot();
+  const {formatMessage: t} = useIntl();
+
+  useEffect(() => {
+    navigation.setOptions({
+      headerLeft: props => (
+        <CustomHeaderLeft
+          tintColor={props.tintColor}
+          headerBackButtonProps={props}
+        />
+      ),
+    });
+  }, [navigation]);
```
this block of code is unnecessary. By default it has a `headerLeft`

cimigree replied:
As I said above, without it, when I go to record another audio recording, there is no header and I am stuck on the screen. Is there another way to make sure there is a header on the page in that situation?

ErikSin commented on src/frontend/screens/Audio/ContentWithControls.tsx:
Everyrtime you use `ContentWithControl`, the prop `controls` uses `<Controls.Row >`. Would it make sense to just integrate `Controls.Row` directly into this component so the developer does not need to import and wrap it everytime?

ErikSin commented on src/frontend/screens/Audio/CreateRecording/RecordingActive.tsx:
```diff
@@ -0,0 +1,64 @@
+
+  useEffect(() => {
+    navigation.setOptions({headerLeft: () => null});
+  }, [navigation]);
```
I think `headerShown:false` would be clearer here.

ErikSin commented on src/frontend/screens/Audio/CreateRecording/useAudioRecording.ts:
```diff
@@ -0,0 +1,117 @@
+          Audio.RecordingOptionsPresets.HIGH_QUALITY,
+          status => {
+            setState(prev => {
+              if (!prev) return prev;
```
what is the logic here? If Im understanding it, it the state is null, don't set the state. But the state is null by default. So we want to ignore the state being set until the recording is complete (which makes sense). But then why the return statement below this line? would we not want to just set the state to null if there is a previous state?

cimigree committed:
Changes from Erik etc.

cimigree committed:
Adds back header for when returning to record another.

cimigree commented:
@ErikSin Ready for continuation of review. I changed the multi ternary to an if statement. I find it a little more readable. I moved the createdAt to happen when the recording is saved. I had to keep the header in RecordingIdle because it is needed when you go to record another audio recording after you have recorded the first one. Got rid of the unlink function because resetting seems to do the trick upon deletion as you noticed.

ErikSin commented on src/frontend/screens/Audio/CreateRecording/RecordingDone.tsx:
```diff
@@ -0,0 +1,147 @@
+          </Pressable>
+        }
+      />
+      <BottomSheetModal ref={deleteSheetRef} isOpen={isDeleteSheetOpen}>
+        <BottomSheetModalContent
+          icon={<ErrorIcon />}
+          title={t(m.deleteBottomSheetTitle)}
+          description={t(m.deleteBottomSheetDescription)}
+          buttonConfigs={[
+            {
+              dangerous: true,
+              text: t(m.deleteBottomSheetPrimaryButtonText),
+              icon: <DeleteIcon color={WHITE} />,
+              onPress: () => {
+                closeDeleteSheet();
+                onDelete();
+              },
+              variation: 'filled',
+            },
+            {
+              variation: 'outlined',
+              text: t(m.deleteBottomSheetSecondaryButtonText),
+              onPress: () => {
+                closeDeleteSheet();
+              },
+            },
+          ]}
+        />
+      </BottomSheetModal>
+      <RecordingSuccessModal
+        sheetRef={successSheetRef}
+        isOpen={isSuccessSheetOpen}
+        onReturnToEditor={handleReturnToEditor}
+        onRecordAnother={handleRecordAnother}
+      />
+    </>
+  );
```
I think it would be better to have one bottom sheet modal with 2 different `< BottomSheetContents/ >` .That way you wouldn't have to import useBottomSheetModalTwice

cimigree replied:
I did it but I personally I don't like it, tbh... Much more messy and seems like a lot more code.

ErikSin commented on src/frontend/screens/ObservationCreate/index.tsx:
```diff
@@ -273,7 +287,9 @@ export const ObservationCreate = ({
         <SaveButton
           onPress={checkAccuracyAndLocation}
           isLoading={
-            createObservationMutation.isPending || createBlobMutation.isPending
+            createObservationMutation.isPending ||
+            createBlobMutation.isPending ||
+            createAudioBlobMutation.isPending
```
`createAudioBlobMutation.isPending` needs to be added to the useEffect dependencies

ErikSin commented on src/frontend/screens/Audio/CreateRecording/useAudioRecording.ts:
```diff
@@ -0,0 +1,35 @@
+    if (!recording) return;
+    await recording.stopAndUnloadAsync();
+    setUri(recording.getURI());
+  }
```
We should probably make all 3 of these function `useCallback` functions.

ErikSin commented on src/frontend/hooks/useDraftObservation.ts:
```diff
@@ -92,6 +94,13 @@ export const useDraftObservation = () => {
+    },
+    [addAudioRecording],
+  );
+
```
creating this function is unneccessary, you can just export `AddAudioRecording` directly (i dont think renaming it is necessary)

ErikSin commented on src/frontend/screens/Audio/Controls.tsx:
```diff
@@ -0,0 +1,119 @@
+        pressed && styles.pressablePressed,
+      ]}>
+      {children}
+    </Pressable>
```
it would probably be good to turn this into a touchable opacity to remain consistent with the rest of the app (unless it drastically changes how it works/looks)

ErikSin commented on src/frontend/screens/Audio/CreateRecording/useAutoStopRecording.ts:
```diff
@@ -0,0 +1,12 @@
+  onPressStop: () => void,
+) {
+  useEffect(() => {
+    if (minutesRemaining === 0) {
```
just to be safe, i think we should do `minutesRemaining <= 0`

ErikSin commented on src/frontend/screens/Audio/Playback.tsx:
this should probably be a shared component because it is going to be used by other screens in the future

ErikSin commented on src/frontend/screens/Audio/constants.ts:
To remain consistent with other constants, i don't think it needs it own file and can just be stored in `Audio.index.tsx`

ErikSin commented on src/frontend/screens/Audio/useAudioPlayback.ts:
```diff
@@ -0,0 +1,85 @@
+  const [isPlaying, setPlaying] = useState(false);
+  const [duration, setDuration] = useState<number>(0);
+  const [currentPosition, setCurrentPosition] = useState(0);
+  const [isReady, setReady] = useState(false);
```
I think we can take out isReady completely (or don't make it state). Its not really being used in a meaningful way, and we don't export it

ErikSin commented on src/frontend/screens/Audio/useAudioPlayback.ts:
```diff
@@ -0,0 +1,85 @@
+  useEffect(() => {
+    let soundInstance: Sound | null = null;
+    Audio.Sound.createAsync({uri: recordingUri})
+      .then(({sound, status}) => {
+        if ('error' in status && status.error) {
+          console.error('Error while creating audio playback', status.error);
+          return;
+        }
```
as a follow up pr, i think we should handle audio errors (because we do not handle them in the UI at all)

cimigree replied:
Ok

ErikSin reviewed (changes requested):
Some adjustment but the overall architecture seems sound. I am a little worried though about the fact that there is no error handling, so I think we should address that in a follow up PR.

cimigree committed:
Wraps everything in row. Prevents going back. Removes header.Adds two modals to the RecordingDone component.

cimigree committed:
Use callback added. Changes to touchable opacity. Adjusts constant location. Adjusts playback location. Removes isready in state.

ErikSin commented on src/frontend/screens/Audio/CreateRecording/index.tsx:
```diff
@@ -1,40 +1,64 @@
+  const {startRecording, stopRecording, reset, status, uri} =
+    useAudioRecording();
+
+  let currentState = 'idle';
+  if (status && status.isRecording) {
+    currentState = 'active';
+  } else if (uri) {
+    currentState = 'done';
+  }
```
slight optimization would be: const currentState = status?.isRecording ? 'active' : uri ? 'done' : 'idle';

ErikSin commented on src/frontend/screens/Audio/CreateRecording/index.tsx:
```diff
@@ -1,40 +1,64 @@
+    );
+  }
+
+  return null;
```
this can be removed

ErikSin commented on src/frontend/screens/Audio/useAudioPlayback.ts:
```diff
@@ -0,0 +1,82 @@
+    };
+  }, [recordingUri, audioCallbackHandler]);
+
+  const startPlayback = async () => {
+    if (!recordedSoundRef.current || isPlaying) return;
+
+    try {
+      if (currentPosition >= duration) {
+        await recordedSoundRef.current!.setPositionAsync(0);
+        setCurrentPosition(0);
+      }
+
+      await recordedSoundRef.current!.playAsync();
+      setPlaying(true);
+    } catch (error) {
+      console.error('Failed to play sound:', error);
+    }
+  };
+
+  const stopPlayback = async () => {
+    if (!recordedSoundRef.current || !isPlaying) return;
+
+    try {
+      await recordedSoundRef.current!.pauseAsync();
+      setPlaying(false);
+    } catch (error) {
+      console.error('Failed to pause sound:', error);
+    }
+  };
```
Sorry, I did not catch this in my last review, but these should also be memoized with useCallback. Generally I think it is good practice to memoize any function or values being exported from a hook

ErikSin commented on src/frontend/screens/Audio/CreateRecording/RecordingDone.tsx:
```diff
@@ -0,0 +1,241 @@
+      );
+    } else if (modalContentType === 'success') {
+      return (
+        <View style={styles.container}>
+          <View style={styles.wrapper}>
+            <SuccessIcon />
+            <Text style={styles.title}>{t(m.successTitle)}</Text>
+            <Text style={styles.description}>
+              <FormattedMessage
+                {...m.successDescription}
+                values={{
+                  audioRecording: t(m.audioRecording),
+                  bold: message => (
+                    <Text style={styles.textBold}>{message}</Text>
+                  ),
+                }}
+              />
+            </Text>
+          </View>
+          <View style={styles.buttonContainer}>
+            <Button
+              fullWidth
+              onPress={handleReturnToEditor}
+              variant="outlined"
+              color="ComapeoBlue">
+              {t(m.returnToEditorButtonText)}
+            </Button>
+            <Button fullWidth onPress={handleRecordAnother}>
+              {t(m.recordAnotherButtonText)}
+            </Button>
+          </View>
+        </View>
```
Is there a reason that `<BottomSheetModalContent />` is not being used here?

ErikSin reviewed (changes requested):
when i delete a recording, the bottom sheet is not closing

cimigree commented:
> when i delete a recording, the bottom sheet is not closing

Hm. I can't figure out why that would be, can't replicate.

Here is what it looks like on my Pixel. Is this what you are talking about?

https://github.com/user-attachments/assets/9de7f5d2-9041-40e4-9be7-551f5615db4f

cimigree committed:
Uses use callback. Changes if to ternary. Uses BottomModalContent.

ErikSin commented:
> > when i delete a recording, the bottom sheet is not closing
> 
> Hm. I can't figure out why that would be, can't replicate.
> 
> Here is what it looks like on my Pixel. Is this what you are talking about?
> deletionvideo.mp4

yeah, i can recreate it sometimes....im just trying to figure out why its only happening sometimes

ErikSin commented:
> > > when i delete a recording, the bottom sheet is not closing
> > 
> > 
> > Hm. I can't figure out why that would be, can't replicate.
> > Here is what it looks like on my Pixel. Is this what you are talking about?
> > deletionvideo.mp4
> 
> yeah, i can recreate it sometimes....im just trying to figure out why its only happening sometimes

hmm, i think i figured how when (but not why) it is happening. If you close the app while taking an observation, and then open the app again, it will open on the `ObservationCreate` screen - which is expected behaviour. But from here if you take an audio clip and try to delete it the bottom sheet will stay open

ErikSin commented:
https://github.com/user-attachments/assets/7a4e3e3e-e46e-4a0e-8901-6b98482c660c

cimigree committed:
Trying to handle dismissing modal when app is closed and restarted.

cimigree commented:
> > > > when i delete a recording, the bottom sheet is not closing
> > > 
> > > 
> > > Hm. I can't figure out why that would be, can't replicate.
> > > Here is what it looks like on my Pixel. Is this what you are talking about?
> > > deletionvideo.mp4
> > 
> > 
> > yeah, i can recreate it sometimes....im just trying to figure out why its only happening sometimes
> 
> hmm, i think i figured how when (but not why) it is happening. If you close the app while taking an observation, and then open the app again, it will open on the `ObservationCreate` screen - which is expected behaviour. But from here if you take an audio clip and try to delete it the bottom sheet will stay open

@ErikSin For me, that is not happening on either of my devices even when I follow the steps you do exactly, unless while in an observation, I actually force close the app and then **restart expo** and then reopen the app. And in that case, that lack of ability to close the sheet doesn't even always happen... I noticed you have the expo development screen when you restart, which I do not. I am thinking this might not happen in the real world because expo won't be restarting mid go?
And because of this I sort of timeboxed this. I worked for 1.5 hours and this is what I came up with and I don't think it really works for the particular situation where expo reloads ... If you know something else to do, I would love to hear it. Or I can completely redo the bottomsheetmodal but I have a feeling that is not a good idea.

ErikSin commented:
Could you build an APK and confirm your theory? :)

ErikSin committed:
chore: update icon on error bottom sheet

ErikSin committed:
chore: added ability for custom message

cimigree committed:
Undoing recent changes

cimigree committed:
Adjusts function definitions and calls at runtime

cimigree committed:
Passing on dismiss

cimigree committed:
Removing go back

cimigree committed:
Trying removal of back handler

cimigree committed:
Trying some fixes

cimigree committed:
Testing out sheet sizes

cimigree committed:
Trying full size

cimigree committed:
Messing with full screen.

cimigree committed:
Trying different order

cimigree committed:
Blocking touch events when modal open.

cimigree committed:
Adds flex.

cimigree committed:
Disabling audio press when modal is open.

cimigree committed:
Timeout

cimigree committed:
Trying without timeout

cimigree committed:
Removes references to isOpen

cimigree committed:
Tries using on dismiss

cimigree committed:
Removes console log. Adds switch statement. Tries timeout with permissons.

cimigree committed:
Uses on dismiss in permissions modal

ErikSin commented on src/frontend/screens/Audio/CreateRecording/index.tsx:
```diff
@@ -1,40 +1,56 @@
+      <RecordingDone
+        uri={uri || ''}
+        duration={status?.durationMillis || 0}
+        onDelete={() => {
+          reset();
+        }}
+        onRecordAnother={reset}
```
onDelete and onReset are now the same thing. It would just be clearer is reset was just a prop of RecordingDone and just pass reset once

ErikSin commented on src/frontend/screens/Audio/CreateRecording/RecordingDone.tsx:
```diff
@@ -0,0 +1,235 @@
+        </View>
+      );
+      return (
+        <View style={{marginTop: 80}}>
```
Also instead of wrapping this, you can just add a `marginTop` of 80 directly to the `SuccessIcon`

ErikSin committed:
chore: updated styling

cimigree committed:
Uses blur event instead of the onDismiss

cimigree committed:
Merge branch 'chore-update-error-icon' of https://github.com/digidem/comapeo-mobile into feat/add-audio-save-playback

cimigree committed:
Passes reset. Removes wrapping.

cimigree committed:
Goes back to using onmodal dismiss

cimigree committed:
Merge branch 'develop' into feat/add-audio-save-playback

cimigree committed:
Fixes mistakes in handling modal dismissal

cimigree committed:
Tidies up modal dismissal action

cimigree committed:
Removes unused prop.

ErikSin commented on src/frontend/screens/Audio/CreateRecording/index.tsx:
```diff
@@ -1,40 +1,51 @@
+  const {startRecording, stopRecording, reset, status, uri} =
+    useAudioRecording();
+
+  const currentState = status?.isRecording ? 'active' : uri ? 'done' : 'idle';
```
nit: can you give this a more meaningful name, perhaps `recordingState`?

ErikSin commented on src/frontend/screens/Audio/PermissionAudioBottomSheetContent.tsx:
```diff
@@ -37,48 +36,38 @@ const m = defineMessages({
 interface PermissionAudioBottomSheetContentProps {
   closeSheet: () => void;
-  navigation: ObservationCreateNavigationProp;
+  setPendingAction: (action: 'navigateToAudio' | null) => void;
```
i think it would be better to pass a function that handles the entire action (since there is only one action that we want), instead of passing the entire setter. Also see comment below about changing the setState to handle a boolean instead of the type `'navigateToAudo' | null`

eg.
instead of `setPendingAction: (action: 'navigateToAudio' | null) => void` do `setNavigateToAudioTrue: ()=>void`

and then in the parent component
```ts
<PermissionBottomSheetContent setNavigateToAudioTrue={()=>{setShouldNavigateToAudio(true)}}  {...rest} />
```

ErikSin commented on src/frontend/screens/Audio/index.tsx:
```diff
@@ -6,6 +6,8 @@ import {CustomHeaderLeft} from '../../sharedComponents/CustomHeaderLeft';
 import {StatusBar} from 'expo-status-bar';
```
this `StatusBar` is causing a flicker in the screen when you navigate back. Notice in the video how the screen jumps up and down when navigating between the screen. I think we can just get rid of it completely. I also think that `Audio/CreateRecording/index` should become the new `Audio/index`

https://github.com/user-attachments/assets/7135013d-a4d6-4975-a015-2793dcaf092a

cimigree replied:
@ErikSin, There is going to be something else in Audio/index in the next PR related to Playback from the Observation. Is it ok if I leave it as is? Not the Status Bar... Having the two files

ErikSin commented on src/frontend/sharedComponents/ActionRow.tsx:
```diff
@@ -51,6 +51,10 @@ export const ActionsRow = ({fieldRefs}: ActionButtonsProps) => {
     openOnMount: false,
   });
 
+  const [pendingAction, setPendingAction] = useState<'navigateToAudio' | null>(
+    null,
+  );
```
I think it is clearer to use a boolean instead of creating 2 opposing states. Eg `const [shouldNavigateToAudio, setShouldNavigateToAudio] = useState(false)`

ErikSin reviewed (changes requested):
Sorry, i just noticed the use of `statusBar`. the SWM team was using them, and it was causing a flickering, so I think we should just remove them entirely

cimigree committed:
Updates var name, removes status bar, changes state for permission.

ErikSin reviewed (approved):
As a follow up, lets add the metadata to the blob
